### PR TITLE
[mrg] Per spec configuration

### PR DESCRIPTION
**Pull request by choldgraf**
In https://github.com/jupyterhub/binderhub/issues/887 we talked about how it'd be cool to have per-repo configuration. This is a first attempt at implementing it.

The basic idea is that we operate similar to how "banned" and "quota increase" functionality works, with one difference:

Instead of a **list** of specification regexes to match to "quota increase" or "banned", we have a **dictionary** where keys are specification regexes, and values are dictionaries of "key:value" pairs that can over-ride configuration on a repo-specific basis.

For example, you could do something like:

```
config:
    GitHubRepoProvider:
      # Add banned repositories to the list below
      # They should be strings that will match "^<org-name>/<repo-name>.*"
      spec_configuration_override:
        ^ines/spacy-binder.*:
		    quota: 1337
        ^binder-examples/requirements.*:
   			banned: true     
```

## Questions to answer

1. I agree w/ @betatim that "spec_configuration_override" isn't great, any suggestions for other names?
2. What's a clean way to document what key:value configuration pairs are meaningful in this case? We'll need to add checks for them one-by-one in the code, and for now I guess we'll just start with a `quota: <int>` key/value. In the future, should we just expose this with good documentation for what the accepted key/values are? (we could also imagine some key/values being repoprovider-specific)

## Need to add

- [x] Docs
- [x] Tests
- [x] Better naming

### CODE REVIEW

File: binderhub/builder.py [hunk 1]
```diff
@@ -424,9 +424,14 @@ def escape(s):
 
     async def launch(self, kube, provider):
         """Ask JupyterHub to launch the image."""
+        # Load the spec-specific configuration if it has been overridden
+        spec_config = provider.spec_configuration_override
+
         # check quota first
         if provider.has_higher_quota():
```
Comment (betatim) [inline binderhub/builder.py]: Should we retire this at the same time as we introduce the new functionality?

File: binderhub/builder.py [hunk 2]
```diff
@@ -424,9 +424,14 @@ def escape(s):
...
             quota = self.settings.get('per_repo_quota_higher')
+        elif "per_repo_quota" in spec_config:
+            quota = spec_config.get('per_repo_quota')
```
Comment (betatim) [inline binderhub/builder.py]:
```suggestion
            quota = spec_config.get('quota')
```
(also on the line above)

Though you could write this as:

```python
quota = spec_config.get('quota', self.settings.get('per_repo_quota'))
```
which is a well used pattern in Python for saying "get this from the dictionary and if it doesn't exist use this thing as default"

File: binderhub/repoproviders.py [hunk 3]
```diff
@@ -111,6 +121,21 @@ def has_higher_quota(self):
                 return True
         return False
 
+    def spec_configuration_override(self):
+        """
+        Return the spec configuration overrides if the given spec has them.
+        """
+        if not isinstance(self.spec_configuration_override, dict):
```
Comment (betatim) [inline binderhub/repoproviders.py]: I think because you [specified `Dict` for the option](https://github.com/jupyterhub/binderhub/pull/888/files#diff-c5688934f1e6dc3e932b6c84c1bbbd5dR83) we should never get here/have to check this as traitlets checks for us.

**Comment by betatim** [comment]
How about `spec_config` as name? With a comment in the doc string of the traitlet to say that you only have to use this to override settings per spec/if no entry is found defaults are used.

For documentation of which keys/options are valid I'd put a pointer in the traitlets' doc that says "repo providers are free to define their own options, see each repo provider for a list of valid keys and their meaning". Then add appropriate docs to the repo providers/base repo provider.

File: binderhub/repoproviders.py [hunk 4]
```diff
@@ -111,6 +121,21 @@ def has_higher_quota(self):
                 return True
         return False
 
+    def spec_configuration_override(self):
```
Comment (betatim) [inline binderhub/repoproviders.py]: What do you think of naming this `spec_config()` and taking care of finding and populating all the default values in this function? Then in the builder we wouldn't need to have any checks for "do we override this or not, what is the default". We have one place that is already ready to use.

The reason to call it `spec_config` is that it would be for one particular spec and not a whole bunch any more. Maybe we could even call it `repo_config`? Depends a bit on what words the builder uses (does it refer to building a spec or a repo). But this might be getting a bit OCD on naming :)

TL;DR: should we compute the configuration (combining over rides and defaults) for the spec here?

File: binderhub/builder.py (hunk 1 above)
Comment (minrk) [reply binderhub/builder.py]: Probably good to have a deprecation transition, in which case include a warning if the old way is used.

File: binderhub/repoproviders.py [hunk 5]
```diff
@@ -111,6 +121,21 @@ def has_higher_quota(self):
...
+        Return the spec configuration overrides if the given spec has them.
+        """
+        if not isinstance(self.spec_configuration_override, dict):
+            raise ValueError(
+                "Specification-specific override is not a "
+                "dictionary, found type %s" % type(self.spec_configuration_override))
+        for spec, config in self.spec_configuration_override:
+            # Ignore case, because most git providers do not
+            # count DS-100/textbook as different from ds-100/textbook
+            if re.match(spec, self.spec, re.IGNORECASE):
```
Comment (minrk) [inline binderhub/repoproviders.py]: FWIW, using `re.match` compiles the regex on each call, which means we are compiling every regex here on every repo provider. Compiling them on application load would make this a bit more efficient.

Related to this: dicts are in a random order, but overrides might want to have a priority if more than one pattern matches the repo. With a dict, the behavior is undefined, with a list it would be consistent. So perhaps a list of tuples is more appropriate than a dict?

File: binderhub/builder.py (hunk 1 above)
Comment (choldgraf) [reply binderhub/builder.py]: I'm fine w/ a transition (though we've still only ever had one release of BinderHub lol)

File: binderhub/repoproviders.py (hunk 3 above)
Comment (choldgraf) [reply binderhub/repoproviders.py]: ah ok - so it'll raise an error if somebody passes a list and not a dict in the helm config?

File: binderhub/repoproviders.py (hunk 4 above)
Comment (choldgraf) [reply binderhub/repoproviders.py]: I like it - that seems reasonable to me, lemme give it a stab and see how it looks

File: binderhub/repoproviders.py (hunk 5 above)
Comment (choldgraf) [reply binderhub/repoproviders.py]: mmmm, I am not sure how to do this :-) feel free to suggest the code that would still allow for config patterns like the one in https://github.com/jupyterhub/binderhub/pull/888#issue-294300289 and I'll give it a shot

File: binderhub/repoproviders.py (hunk 4 above)
Comment (choldgraf) [reply binderhub/repoproviders.py]: see the latest push for an attempt at what this would look like, is this the kinda thing you had in mind? 

https://github.com/jupyterhub/binderhub/pull/888/commits/1399d58986da19c77139b8a9b8cedadc25c13c91#diff-c5688934f1e6dc3e932b6c84c1bbbd5dR133

In this case, `spec_config` would be defined at the helm chart config level, and repo_config is a method of RepoProvider that returns a dictionary of configuration values for that *repository* (which might have been updated from the spec_config setting)

File: binderhub/builder.py (hunk 1 above)
Comment (betatim) [reply binderhub/builder.py]: True on paper and in practice we release several times a week and people actually use BinderHub so I think we should be nice to people. The "higher quota" thing has only existed for a week or two so :-/

```python
warnings.warn(
    "XXX is deprecated, use YYY instead",
    DeprecationWarning
)
```
is the generic Python warnings thing. It isn't ideal (most people won't see deprecation warnings except in their tests). Does traitlets have a tool for deprecations?

File: binderhub/repoproviders.py (hunk 5 above)
Comment (betatim) [reply binderhub/repoproviders.py]: Good point about the random order! I had pondered that one might have multiple patterns and if they should be additive or "last one wins" or what, then decided that for a first pass we should ignore that. However the random order thing is something we need to address (or inform the user that they have more than one pattern that matches and that this is an error (for now)).

I think what we want is a list like:
```python
specs = [{"pattern": "some-pattern", "quota": 10}, {"pattern": "some-other-pattern", "quota": 12}]
```
in YAML it would look like:
```yaml
# not super sure about the first one but the second two should work
- {pattern: "yet another pattern", quota: 33}
- pattern: some-pattern
  quota: 10
- pattern: some-other-pattern
  quota: 12
```

Comment (choldgraf) [reply binderhub/repoproviders.py]: Ah I like this pattern! @minrk does it look good to you as well?

File: binderhub/builder.py (hunk 1 above)
Comment (minrk) [reply binderhub/builder.py]: Due to the invisible-by-default deprecation warnings, I think that application-level deprecation warnings should actually *not* use DeprecationWarning, so they are visible by default. It's a tricky choice since deprecation warnings are always safely ignored "for now" but if folks never see them, they may as well not exist. As for a transition, I was mostly thinking of ourselves at mybinder.org, since it would allow us to decouple "adopt the new way" from "bump binderhub" PRs without config changes.

FWIW, I don't think the deprecation needs to be a blocker here, as the old way still works. Cleanly deprecating the old way can be a follow-up PR.

> Does traitlets have a tool for deprecations?

Traitlets doesn't have any helpers for deprecations, but the way traitlets works can make writing deprecations handy since deprecations can go in observers, and completely removed from the bodies of methods, etc. for example:

```python
@observe('old_way')
def _old_way_changed(self, change):
    warnings.warn("old way is deprecated as of...new way is...")
    self.new_way = modified_form(change.new)
    # never need to look directly at self.old_way after this

...

def method(self):
    only_use(self.new_way)
```

The simplest version of this can lead to undefined behavior (which wins?) if both the old and new way are specified in the same config, but I think that's not a big deal if the old-way warning is visible.

File: binderhub/repoproviders.py (hunk 5 above)
Comment (minrk) [reply binderhub/repoproviders.py]: yes, that looks great to me! The only question is that 'pattern' is in the same namespace as the config overrides. This is nice since it's more concise, but could allow collisions or confusion. The more rigorous, but slightly more tedious would be to explicitly separate the match from the override config:

```yaml
- pattern: some-pattern
  config:
    quota: 10
    other_config: "x"
```

This would allow adding other siblings to `pattern` if we e.g. had some other options that should influence the selection rather than the overridden config. I don't have strong feelings, but I've occasionally regretted not separating things like this in JupyterHub in the past (see spawner options in the REST API).

You can see the difference in code where 'pattern' needs to be handled specially:

```python
# copy because we need to modify the dict
update_config = config.copy()
# remove pattern before updating to config because pattern is not part of the config:
pattern = config.pop('pattern')
if pattern.match(...):
    config.update(update_config)
```

vs:

```python
pattern = item['pattern']
update_config = item['config']
if pattern.match(spec):
    config.update(update_config)
```

File: binderhub/builder.py (hunk 1 above)
Comment (betatim) [reply binderhub/builder.py]: I think using the traitlets `observe` together with printing a warning (via `warnings.warn()`) is a good way to do this.

**Comment by betatim** [comment]
> FWIW, I don't think the deprecation needs to be a blocker here, as the old way still works. Cleanly deprecating the old way can be a follow-up PR.

That is a good point. Let's follow that suggestion. Being able to bump binderhub on mybinder.org without having to also change the config makes life easier (and lets us use henchbot!).

**Comment by choldgraf** [comment]
See the latest push for the following updates:

- The API should now follow what @minrk suggested, how does that look?
- A buncha tests added (let's see if they pass!)
- Some [minor docs added](https://369-89419368-gh.circle-artifacts.com/0/html/customizing.html#custom-configuration-for-specific-repositories)

File: binderhub/repoproviders.py [hunk 6]
```diff
@@ -111,6 +126,38 @@ def has_higher_quota(self):
                 return True
         return False
 
+    def repo_config(self, settings):
```
Comment (choldgraf) [inline binderhub/repoproviders.py]: should we make this a property or keep an explicit method?

Comment (minrk) [reply binderhub/repoproviders.py]: I think it can't be a property if it's taking arguments.

File: binderhub/repoproviders.py [hunk 7]
```diff
@@ -111,6 +126,38 @@ def has_higher_quota(self):
...
+
+        # Defaults and simple overrides
+        if self.has_higher_quota():
+            repo_config['quota'] = settings.get('per_repo_quota_higher')
+        else:
+            repo_config['quota'] = settings.get('per_repo_quota')
+
+        # Spec regex-based configuration
+        for item in self.spec_config:
+            pattern = item.pop('pattern', None)
```
Comment (minrk) [inline binderhub/repoproviders.py]: I think we want `get` not `pop`, to avoid modifying the item.

File: binderhub/tests/test_repoproviders.py [hunk 8]
```diff
@@ -92,6 +92,65 @@ def test_higher_quota():
...
+        "pattern": '^yuvipanda.*',
+        "config": {
+            "key1": "val1",
+            "quota": 999
+        }
+    }
+
+    provider = GitHubRepoProvider(
+        spec='jupyterhub/zero-to-jupyterhub-k8s/v0.4',
+        config=[base_config]
```
Comment (minrk) [inline binderhub/tests/test_repoproviders.py]: spec_config?

**Comment by betatim** [comment]
I'd got with option (1) for the moment in terms of priority/when to stop.

File: binderhub/repoproviders.py (hunk 6 above)
Comment (betatim) [reply binderhub/repoproviders.py]: I am happy with it being a method.

**Comment by choldgraf** [comment]
thanks for the feedback on tests etc

The latest commit:

* Fixes the tests so they pass
* Adds a test for default behavior w/o a matching spec
* Adds a note to documentation about configuration override

**Comment by betatim** [comment]
Nice work! Tests, docs and a new feature!

---

What is the etiquette for marking conversations as "resolved"? Can I do it in a PR that isn't mine? Should the author do it? When I am the author I tend to mark things are solved (if they don't auto collapse) when I have implemented the feedback or the discussion has somehow ended. In this PR there were several people involved so it felt weird to do that but I also found myself scrolling through each discussion several times trying to not miss new comments.

**Comment by choldgraf** [comment]
Re: "resolved" that's a good question, I'm generally happy for people to resolve things even if I'm the author on the PR, as it often means that my to-do list has shrunk :-)

as with the other team compass guidelines, I trust the judgment of other folks on the project so I'm happy for them to take the initiative there
